### DISCUSSION TITLE
[Form] Add support to control with other primitives

### DISCUSSION THREAD

**Discussion by martinxu9** [radix-ui/primitives#3950] [Ideas]
## Question

Regarding Forms, is there a timeline on the below gap (quoting from [here](https://www.radix-ui.com/primitives/docs/components/form#examples) )? Thanks! 

> Note: At the moment, it is not possible to compose Form with Radix's other form primitives such as Checkbox, Select, etc. We are working on a solution for this.

**Comment by benoitgrelard** [comment]
This is something we definitely want to look at but haven't had the time to prioritize.
It will require significant changes under the hood for these primitives, so it isn't a small change.

**Comment by jd-carroll** [comment]
Let's say I had a few days to invest in getting radix/form working with radix/themes (not that I expect I would produce releasable quality code).

The core issue blocking radix/primitives from working with radix/form is not immediately apparent to me. Is it an issue with the context's?

What I'd like to do is have an SSR form that is displayed when the user first loads the site (think sign-up page). This seems pretty straight forward (even with radix/themes) using the strategy already laid out. (ie. leverage defined validity "states" with defined messages and then `data-` states and `onChange` handlers on the input component)

But I'm sure I'm missing some significant pieces here, mostly just curious.

**Comment by bdrtsky** [comment]
@jd-carroll I did some small research, the issue with Radix primitives is that it seems like `onInvalid` event is not propagated, since Radix rendering other elements as `button` etc, and actual HTML `input` or `select` is hidden. On the form you can listen `onInvalid` with Radix primitives just fine! I feel like you need to pass those events to primitives Root and this should solve the issue.

**Comment by bdrtsky** [comment]
Another thing you would need to do, is to handle focus properly, because currently Form trying to focus aria-hidden inputs on validation, which throws an error in devtools
